---
title: "Post-colocalization data exploration"
subtitle: "Format colocalization results for further analysis and visualize results"
output: 
  html_document:
    toc: true
    toc_depth: 2
author: "Mike Gloudemans"
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r include=FALSE}
require(reshape2)
require(ggplot2)
require(dplyr)
require(readr)
```

# Format colocalization results table

For this script, we assume that the accompanying script `summarize_pre_test_statistics.Rmd`
has already been run for the colocalization run of interest.

These results have been filtered to include only the relevant GWAS traits and to remove
any colocalization runs that didn't pass significance after intersection with the eQTL
summary statistics.

### Load filtered and QC'ed COLOC results from the previous script.

```{r}
results = read.csv("/users/mgloud/projects/insulin_resistance/eriks_bonus/output/full_coloc_results_qced.txt", sep=",")
results$chr = sapply(as.character(results$ref_snp), function(x) {strsplit(x, "_")[[1]][1]})
results$pos = sapply(as.character(results$ref_snp), function(x) {strsplit(x, "_")[[1]][2]})
```

We collect and add in the necessary metadata to fill out the specified
data output format.

```{r include}
# Get HGNC gene names
genes = read.table("/users/mgloud/projects/insulin_resistance/eriks_grant/mart_export.txt", header=TRUE, sep="\t")
genes = genes[,-2]
names(genes) = c("ensembl", "hgnc")
genes = genes[!duplicated(genes$ensembl),]
results = merge(genes, results, all.y=TRUE)

# Load rsIDs -- here, we'll use a simple lookup in the 1Kgenomes file
rsids = results[!duplicated(results$ref_snp),][c("ref_snp", "chr", "pos")]
rsids$rsid = sapply(1:dim(rsids)[1], function(i)
	{
		chr = rsids$chr[i]
		pos = rsids$pos[i]

		query = system(paste0("tabix /mnt/lab_data/montgomery/shared/1KG/ALL.chr", chr, ".phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz ", chr, ":", pos, "-", pos), intern=TRUE)
		if (length(query) > 1)
		{
			vars = sapply(1:length(query), function(i) {strsplit(query[[i]], "\\t")[[1]][3]})
			vars2 = vars[grepl("rs", vars)]
			if (length(vars2) > 0)
			{
				return(paste(vars2, collapse="_"))
			}
			return(paste(vars, collapse="_"))
		}
		if (length(query) == 0)
		{
			print("No variant found at this locus. Please check the code for errors.")
			return(rsids$ref_snp[i])
		}
		rsid = strsplit(query, "\\t")[[1]][3]
		return(rsid)
	}
)
results = merge(rsids[c("ref_snp", "rsid")], results, by="ref_snp")

ensembl_locs = read_delim("/srv/scratch/restricted/rare_diseases/data/annot/gencode.v19.annotation.genes.bed.gz", delim="\t", col_names=FALSE)
colnames(ensembl_locs) = c("chrom", "gene_start", "gene_end", "feature")

results$gene_start = ensembl_locs[match(results$feature, ensembl_locs$feature),]$gene_start
results$gene_end = ensembl_locs[match(results$feature, ensembl_locs$feature),]$gene_end


results$min_gwas_pval = 10^-(results$X.log_gwas_pval)
results$min_eqtl_pval = 10^-(results$X.log_eqtl_pval)

```

### View the assignment of SNPs to loci

We can verify that any group of SNPs within less than 1MB of one another have
been assigned the same locus number.

```{r}
# Make sure the assigned locus numbers are reasonable.
stopifnot(length(unique(results$locus)) == max(results$locus))

# View the assignment of SNPs to loci
sub = results[!duplicated(results$ref_snp),]
sub[order(as.character(sub$ref_snp)),][c("ref_snp", "locus")]
```

### Output our results file for further in-depth analysis

```{r}
results_output = results[c("rsid", "chr", "pos", "locus", "min_gwas_pval", "min_eqtl_pval", "gwas_trait", "eqtl_file", "ensembl", "hgnc", "gene_start", "gene_end", "clpp", "clpp_mod", "n_snps", "all_sig_gwas", "all_sig_eqtl")]
write.table(results_output, "/users/mgloud/projects/insulin_resistance/eriks_bonus/output/clpp_results.txt", quote=FALSE, col.names=TRUE, row.names=FALSE, sep="\t")
```

# Compare CLPP and CLPP-mod scores

```{r}
sub = results[results$X.log_gwas_pval, results$X.log_gwas_pval]

# Ranked
plot(rank(results$clpp), rank(results$clpp_mod))
# Ranked filtered
plot(rank(results$clpp), rank(results$clpp_mod))

# Absolute
plot(results$clpp, results$clpp_mod)

# Ranked absolute

```

# Visualize results using heatmaps

```{r}
# Initalize a 3D CLPP array of tissue x gene x gwas_trait
# Maybe initialize another indicator array to show which ones meet all the cutoffs (could then remove slices with no results meeting the cutoffs)
tissues = unique(results$eqtl_file)
genes = unique(results$feature)
gwas_traits = unique(results$gwas_trait)

#min_clpp = 0.001
min_clpp = 0
clpp = array(min_clpp, dim = c(length(tissues), length(genes), length(gwas_traits)))

dimnames(clpp)[[1]] = tissues
dimnames(clpp)[[2]] = genes
dimnames(clpp)[[3]] = gwas_traits

# Fill the array
for (i in 1:dim(clpp)[1])
{
	for (j in 1:dim(clpp)[2])
	{
		for (k in 1:dim(clpp)[3])
		{
			clpp[i,j,k] = max(results[(results$eqtl_file == tissues[i]) & (results$feature == genes[j]) & (results$gwas_trait == gwas_traits[k]),]$clpp_mod, clpp[i,j,k])
		}
	}
}

clpp = clpp[,apply(clpp, 2, max) >= 0.3,]


# Cluster the array within each plane
# May need to read a bit about how to do this and how to choose correlation parameters...but don't waste too much time on it
# TODO: (I'll skip it for now, just focus on consistency)

# Plot it in interesting ways? For now though just do layered heatmaps, one for each tissue
# (Actually, would be nice to be able to view all 3 possible slice orientations)

# Make one plot for each tissue
for (i in 1:dim(clpp)[1])
{
	tissue = tissues[i]

	my_matrix = as.data.frame(as.matrix(clpp[i,,]))
	my_matrix$gene = rownames(my_matrix)

	# Plot heatmap
	heat = melt(my_matrix)
	heat$plot_vals = heat$value
	heat$binned_vals = NA
	heat$binned_vals[heat$plot_vals <= 1] = 8
	heat$binned_vals[heat$plot_vals < 0.9] = 7
	heat$binned_vals[heat$plot_vals < 0.8] = 6
	heat$binned_vals[heat$plot_vals < 0.7] = 5
	heat$binned_vals[heat$plot_vals < 0.6] = 4
	heat$binned_vals[heat$plot_vals < 0.5] = 3
	heat$binned_vals[heat$plot_vals < 0.4] = 2
	heat$binned_vals[heat$plot_vals < 0.3] = 1

	heat$binned_vals = as.factor(heat$binned_vals)
	
	heat$gene = factor(heat$gene, levels=unique(heat$gene))
	heat$variable = factor(heat$variable, levels=unique(heat$variable))

	heat$strike = ""
	heat$strike[is.na(heat$binned_vals)] = "X"
	heat$binned_vals[is.na(heat$binned_vals)] = 1

	g = ggplot(data = heat, aes(x = variable, y = gene)) +
			geom_tile(aes(fill = binned_vals), color="grey80", size=0.5) +
			scale_fill_manual(drop=FALSE, values=colorRampPalette(c("white","red"))(8), name="CLPP_mod", labels = c("0-0.3", "0.3-0.4", "0.4-0.5", "0.5-0.6", "0.6-0.7", "0.7-0.8", "0.8-0.9", "0.9-1"),na.value = "white") +
			theme(axis.text.y=element_text(size=10),
			axis.text.x = element_text(angle = 90, hjust = 1)) +
			labs(y = "Nearest Gene --> Colocalized Gene") +
			labs(x = "GWAS Trait") + 
			geom_text(label=heat$strike, size=4, color="red", alpha=0.4)

	ggsave(paste0("/users/mgloud/projects/insulin_resistance/output/new/insulin_", tissue, "_full_clpp.png"), width = 8, height = 30, units = "in", dpi = 300, limitsize=FALSE)
}


# Make one plot for each gene
for (i in 1:dim(clpp)[2])
{
	gene = genes[i]

	my_matrix = as.data.frame(as.matrix(clpp[,i,]))
	my_matrix$tissue = rownames(my_matrix)

	# Plot heatmap
	heat = melt(my_matrix)
	heat$plot_vals = heat$value
	heat$binned_vals = NA
	heat$binned_vals[heat$plot_vals <= 1] = 8
	heat$binned_vals[heat$plot_vals < 0.9] = 7
	heat$binned_vals[heat$plot_vals < 0.8] = 6
	heat$binned_vals[heat$plot_vals < 0.7] = 5
	heat$binned_vals[heat$plot_vals < 0.6] = 4
	heat$binned_vals[heat$plot_vals < 0.5] = 3
	heat$binned_vals[heat$plot_vals < 0.4] = 2
	heat$binned_vals[heat$plot_vals < 0.3] = 1

	heat$binned_vals = as.factor(heat$binned_vals)
	
	heat$tissue = factor(heat$tissue, levels=unique(heat$tissue))
	heat$variable = factor(heat$variable, levels=unique(heat$variable))

	heat$strike = ""
	heat$strike[is.na(heat$binned_vals)] = "X"
	heat$binned_vals[is.na(heat$binned_vals)] = 1

	g = ggplot(data = heat, aes(x = variable, y = tissue)) +
			geom_tile(aes(fill = binned_vals), color="grey80", size=0.5) +
			scale_fill_manual(drop=FALSE, values=colorRampPalette(c("white","red"))(8), name="CLPP_mod", labels = c("0-0.3", "0.3-0.4", "0.4-0.5", "0.5-0.6", "0.6-0.7", "0.7-0.8", "0.8-0.9", "0.9-1"),na.value = "white") +
			theme(axis.text.y=element_text(size=10),
			axis.text.x = element_text(angle = 90, hjust = 1)) +
			labs(y = "Tissue") +
			labs(x = "GWAS Trait") + 
			geom_text(label=heat$strike, size=4, color="red", alpha=0.4)

	ggsave(paste0("/users/mgloud/projects/insulin_resistance/output/new/insulin_", gene, "_full_clpp.png"), width = 8, height = 30, units = "in", dpi = 300, limitsize=FALSE)
}


# Make one plot for each GWAS
for (i in 1:dim(clpp)[3])
{
	gwas= gwas_traits[i]

	my_matrix = as.data.frame(t(as.matrix(clpp[,,i])))
	my_matrix$gene = rownames(my_matrix)

	# Plot heatmap
	heat = melt(my_matrix)
	heat$plot_vals = heat$value
	heat$binned_vals = NA
	heat$binned_vals[heat$plot_vals <= 1] = 8
	heat$binned_vals[heat$plot_vals < 0.9] = 7
	heat$binned_vals[heat$plot_vals < 0.8] = 6
	heat$binned_vals[heat$plot_vals < 0.7] = 5
	heat$binned_vals[heat$plot_vals < 0.6] = 4
	heat$binned_vals[heat$plot_vals < 0.5] = 3
	heat$binned_vals[heat$plot_vals < 0.4] = 2
	heat$binned_vals[heat$plot_vals < 0.3] = 1

	heat$binned_vals = as.factor(heat$binned_vals)
	
	heat$gene = factor(heat$gene, levels=unique(heat$gene))
	heat$variable = factor(heat$variable, levels=unique(heat$variable))

	heat$strike = ""
	heat$strike[is.na(heat$binned_vals)] = "X"
	heat$binned_vals[is.na(heat$binned_vals)] = 1

	g = ggplot(data = heat, aes(x = variable, y = gene)) +
			geom_tile(aes(fill = binned_vals), color="grey80", size=0.5) +
			scale_fill_manual(drop=FALSE, values=colorRampPalette(c("white","red"))(8), name="CLPP_mod", labels = c("0-0.3", "0.3-0.4", "0.4-0.5", "0.5-0.6", "0.6-0.7", "0.7-0.8", "0.8-0.9", "0.9-1"),na.value = "white") +
			theme(axis.text.y=element_text(size=10),
			axis.text.x = element_text(angle = 90, hjust = 1)) +
			labs(y = "Gene") +
			labs(x = "Tissue") + 
			geom_text(label=heat$strike, size=4, color="red", alpha=0.4)

	ggsave(paste0("/users/mgloud/projects/insulin_resistance/output/new/insulin_", gwas, "_full_clpp.png"), width = 8, height = 30, units = "in", dpi = 300, limitsize=FALSE)
}


stopifnot(FALSE)


# Use 3D heatmap vis? We'll see, this isn't top priority right now.
# A good long-term goal

```
