---
title: "Post-colocalization data exploration"
subtitle: "Format colocalization results for further analysis and visualize results"
output: 
  html_document:
    toc: true
    toc_depth: 2
author: "Mike Gloudemans"
date: "`r format(Sys.time(), '%d %B, %Y')`"
---

```{r include=FALSE}
require(reshape2)
require(ggplot2)
require(dplyr)
require(readr)
require(rjson)
```

# Format colocalization results table

For this script, we assume that the accompanying script `summarize_pre_test_statistics.Rmd`
has already been run for the colocalization run of interest.

These results have been filtered to include only the relevant GWAS traits and to remove
any colocalization runs that didn't pass significance after intersection with the eQTL
summary statistics.

### Load filtered and QC'ed COLOC results from the previous script.

```{r}
# Config file should be specified in the script that renders this file
config = fromJSON(file=config_file)

results = read.table(paste(config$out_dir, "full_coloc_results_qced.txt", sep="/"), sep="\t", header=TRUE)
results$chr = sapply(as.character(results$ref_snp), function(x) {strsplit(x, "_")[[1]][1]})
results$pos = sapply(as.character(results$ref_snp), function(x) {strsplit(x, "_")[[1]][2]})
```

We collect and add in the necessary metadata to fill out the specified
data output format.

```{r include}
# Get HGNC gene names
genes = read.table("/users/mgloud/projects/insulin_resistance/scripts/auxiliary/eriks_bonus/eriks_grant/mart_export.txt", header=TRUE, sep="\t")
genes = genes[,-2]
names(genes) = c("ensembl", "hgnc")
genes = genes[!duplicated(genes$ensembl),]
results = merge(genes, results, all.y=TRUE)

# Load rsIDs -- here, we'll use a simple lookup in the 1Kgenomes file
rsids = results[!duplicated(results$ref_snp),][c("ref_snp", "chr", "pos")]
rsids$rsid = sapply(1:dim(rsids)[1], function(i)
	{
		chr = rsids$chr[i]
		pos = rsids$pos[i]
		query = system(paste0("tabix /mnt/lab_data/montgomery/shared/1KG/hg38/ALL.chr", chr, "_GRCh38.genotypes.20170504.vcf.gz ", chr, ":", pos, "-", pos), intern=TRUE)
		if (length(query) > 1)
		{
			vars = sapply(1:length(query), function(i) {strsplit(query[[i]], "\\t")[[1]][3]})
			vars2 = vars[grepl("rs", vars)]
			if (length(vars2) > 0)
			{
				return(paste(vars2, collapse="_"))
			}
			return(paste(vars, collapse="_"))
		}
		if (length(query) == 0)
		{
			print("No variant found at this locus. Please check the code for errors.")
			return(rsids$ref_snp[i])
		}
		rsid = strsplit(query, "\\t")[[1]][3]
		return(rsid)
	}
)
results = merge(rsids[c("ref_snp", "rsid")], results, by="ref_snp")

# I got rid of this for now because I changed it to be for hg38 instead. Really though, I should adapt this
# code to work for both
#ensembl_locs = read_delim("/srv/scratch/restricted/rare_diseases/data/annot/gencode.v19.annotation.genes.bed.gz", delim="\t", col_names=FALSE)
#colnames(ensembl_locs) = c("chrom", "gene_start", "gene_end", "feature")

#results$gene_start = ensembl_locs[match(results$feature, ensembl_locs$feature),]$gene_start
#results$gene_end = ensembl_locs[match(results$feature, ensembl_locs$feature),]$gene_end


results$min_gwas_pval = 10^-(results$X.log_gwas_pval)
results$min_eqtl_pval = 10^-(results$X.log_eqtl_pval)

```

### View the assignment of SNPs to loci

We can verify that any group of SNPs within less than 1MB of one another have
been assigned the same locus number.

```{r}
# Make sure the assigned locus numbers are reasonable.
stopifnot(length(unique(results$locus)) == max(results$locus))

# View the assignment of SNPs to loci
sub = results[!duplicated(results$ref_snp),]
sub[order(as.character(sub$ref_snp)),][c("ref_snp", "locus")]
```

### Output our results file for further in-depth analysis

```{r}
#results_output = results[c("rsid", "chr", "pos", "locus", "min_gwas_pval", "min_eqtl_pval", "gwas_trait", "eqtl_file", "ensembl", "hgnc", "gene_start", "gene_end", "clpp", "clpp_mod", "n_snps", "all_sig_gwas", "all_sig_eqtl")]
results_output = results[c("rsid", "chr", "pos", "locus", "min_gwas_pval", "min_eqtl_pval", "gwas_short", "eqtl_short", "gwas_trait", "eqtl_file", "feature", "ensembl", "hgnc", "clpp", "clpp_mod", "n_snps", "all_sig_gwas", "all_sig_eqtl", "base_gwas_file")]
write.table(results_output, paste0(config$out_dir, "/clpp_results_", config$analysis_date, ".txt"), quote=FALSE, col.names=TRUE, row.names=FALSE, sep="\t")
```

Note: There used to be additional analysis at the end of this script, but I deleted it because most if not all of it has been superseded
by other, superior analyses. To find it again, I'd have to look on the Github repository.

